src/Surfnet/StepupSelfService/SelfServiceBundle/Service/ActivationFlowService.php

```diff
@@ -42,19 +43,60 @@ class ActivationFlowService
      *
      * Note that:
      * - fieldName and options are configured in the SelfServiceExtension
+     *
+     * @param string[] $options
+     * @param array<string, string> $attributes
      */
     public function __construct(
-        private readonly RequestStack $requestStack,
+        private readonly AuthenticatedSessionStateHandler $sessionState,
+        private readonly TokenStorageInterface $tokenStorage,
         private readonly LoggerInterface $logger,
         private readonly string $fieldName,
-        private readonly array $options
+        private readonly array $options,
+        private readonly string $attributeName,
+        private readonly array $attributes
     ) {
     }
 
-    public function process(string $uri): void
+    public function processPreferenceFromUri(string $uri): void
+    {
+        $requestedActivationPreference = $this->getFlowPreferenceFromUri($uri);
+        if ($requestedActivationPreference instanceof ActivationFlowPreferenceNotExpressed) {
+            return;
+        }
+
+        $this->logger->info('Storing the preference in session');
+        $this->sessionState->setRequestedActivationFlowPreference($requestedActivationPreference);
+    }
+
+    public function getPreference(): ActivationFlowPreferenceInterface
+    {
+        $requestedActivationPreference = $this->sessionState->getRequestedActivationFlowPreference();
+        $availableActivationPreferences = $this->getFlowPreferencesFromSamlAttributes();
+
+        if (count($availableActivationPreferences) == 0) {
+            $this->logger->info('No entitlement attributes found to determine the allowed flow, allowing all flows');
+            $availableActivationPreferences = [
+                'ra', 'self'
```

Comment:
Suggestion: you reference these `ra` and `self` options a couple of times in this service. You might want to make them class constants. Or maybe express them in an `enum`? Also usefull to reference in the ActivationFlowPreference value object.

Reply:
I'm now using the preference VO for this.